clustering/clustering.py

```diff
@@ -0,0 +1,446 @@
+            features (str): name of the features to be used for nearest neighbors computation. 
+                Available features are listed in the clustering settings file.
+
+        Returns:
+            (nx.Graph): NetworkX graph representation of sounds.
+        """
+        # Create k nearest neighbors graph        
+        graph = nx.Graph()
+        graph.add_nodes_from(sound_ids_list)
+        # we set k to log2(N), where N is the number of elements to cluster. This allows us to reach a sufficient number of 
+        # neighbors for small collections, while limiting it for larger collections, which ensures low-computational complexity.
+        k = int(np.ceil(np.log2(len(sound_ids_list))))
+
+        for sound_id in sound_ids_list:
+            try:
+                nearest_neighbors = self.gaia.search_nearest_neighbors(sound_id, k, sound_ids_list, features=features)
+                # edges += [(sound_id, i[0]) for i in nearest_neighbors if i[1]<clust_settings.MAX_NEIGHBORS_DISTANCE]
+                graph.add_edges_from([(sound_id, i[0]) for i in nearest_neighbors if i[1]<clust_settings.MAX_NEIGHBORS_DISTANCE])
+                # graph.add_weighted_edges_from([(sound_id, i[0], 1/i[1]) for i in nearest_neighbors if i[1]<clust_settings.MAX_NEIGHBORS_DISTANCE])
+            except ValueError:  # node does not exist in Gaia dataset
+                graph.remove_node(sound_id)
+
+        # Remove isolated nodes
+        graph.remove_nodes_from(list(nx.isolates(graph)))
+
+        return graph
+
+    def create_common_nn_graph(self, sound_ids_list, features=clust_settings.DEFAULT_FEATURES):
+        """Creates a Common Nearest Neighbors Graph representation of the given sounds.
+
+        Args:
+            sound_ids_list (List[str]): list of sound ids.
+            features (str): name of the features to be used for nearest neighbors computation. 
+                Available features are listed in the clustering settings file.
+
+        Returns:
+            (nx.Graph): NetworkX graph representation of sounds.
+        """
+        # first create a knn graph
+        knn_graph = self.create_knn_graph(sound_ids_list, features=features)
+
+        # create the common nn graph
+        graph = nx.Graph()
+        graph.add_nodes_from(knn_graph.nodes)
+
+        for i, node_i in enumerate(knn_graph.nodes):
+            for j, node_j in enumerate(knn_graph.nodes):
+                if j > i:
+                    num_common_neighbors = len(set(knn_graph.neighbors(node_i)).intersection(knn_graph.neighbors(node_j)))
+                    if num_common_neighbors > 0:
+                        graph.add_edge(node_i, node_j, weight=num_common_neighbors)
+
+        # keep only k most weighted edges
+        k = int(np.ceil(np.log2(len(graph.nodes))))
+        for node in graph.nodes:
+            ordered_neighbors = sorted(list(graph[node].iteritems()), key=lambda x: x[1]['weight'], reverse=True)
```

Comment:
iteritems isn't in python 3. You could just use items (remember that in py2 this makes a copy, so if it's huge this might cause problems. If you really need an iterable version, use `six.iteritems`

Reply:
what's the difference between graph.nodes and graph[node]? Is there a way that allows you to get the name/values in the for loop?